### ISSUE TITLE
`rc_buffer` should not be marked as MachineApplicable

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#6241]
I tried this code (https://github.com/rust-lang/rust/blob/0da6d42f297642a60f2640ec313b879b376b9ad8/compiler/rustc_passes/src/liveness.rs#L262):

```rust
    fn set_captures(&mut self, hir_id: HirId, cs: Vec<CaptureInfo>) {
        self.capture_info_map.insert(hir_id, Rc::new(cs));
    }
```

I expected to see this happen: `rc_buffer` warns, but doesn't change the code since it would also have to change the type of `capture_info_map`.

Instead, this happened: `rc_buffer` tries and fails to substitute the types.

```
error[E0308]: mismatched types
   --> compiler/rustc_passes/src/liveness.rs:262:46
    |
262 |         self.capture_info_map.insert(hir_id, Rc::new(cs));
    |                                              ^^^^^^^^^^^ expected slice, found struct `std::vec::Vec`
    |
    = note: expected struct `std::rc::Rc<[liveness::CaptureInfo]>`
               found struct `std::rc::Rc<std::vec::Vec<liveness::CaptureInfo>>`

warning: usage of `Rc<T>` when T is a buffer type
   --> compiler/rustc_passes/src/liveness.rs:193:32
    |
193 |     capture_info_map: HirIdMap<Rc<Vec<CaptureInfo>>>,
    |                                ^^^^^^^^^^^^^^^^^^^^ help: try: `Rc<[CaptureInfo]>`
    |
    = note: `#[warn(clippy::rc_buffer)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#rc_buffer
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (ffa2e7a 2020-10-24)
